Math/NumberTheory/Primes/Factorisation/QuadraticSieve.hs

```diff
@@ -129,40 +141,74 @@ appearsOnlyOnce = fst . L.foldl' go (mempty, mempty)
 -- By construction, for any given prime, this number is even. From here, a
 -- square root is computed.
-findSecondSquare :: Integer -> [(S.IntSet, PS.PrimeIntSet)] -> S.IntSet -> Integer
-findSecondSquare n indexedFactorisations solution = I.foldrWithKey computeRoot 1 countPowers
+findSecondSquare :: Integer -> V.Vector PS.PrimeIntSet -> [Int] -> Integer
+findSecondSquare n factorisations solution = I.foldrWithKey computeRoot 1 countPowers
   where
-    computeRoot key power previous = (intToInteger key ^ (power `div` 2 :: Int) * previous) `mod` n
+    computeRoot key power previous = (intToInteger key ^ (if even power then (power `div` 2 :: Int) else error "Wrong second square") * previous) `mod` n
     countPowers = foldl count I.empty squares
     count = PS.foldr (\prime im -> I.insertWith (+) (unPrime prime) (1 :: Int) im)
-    squares = fmap snd (filter (\(index, _) -> index `S.isSubsetOf` solution) indexedFactorisations)
+    squares = map (factorisations V.!) solution
+
+data SomeKnown (f :: Nat -> Type) where
+  SomeKnown :: KnownNat k => f k -> SomeKnown f
+
+convertToList :: SomeKnown DBVector -> [Int]
+convertToList (SomeKnown solution) = listBits $ SU.fromSized $ unDBVector solution
+
+translate :: [SignedPrimeIntSet] -> SomeKnown SBMatrix
+translate listOfFactorisations = translateHelper listOfFactorisations (length listOfFactorisations)
+  where
+    translateHelper :: [SignedPrimeIntSet] -> Int -> SomeKnown SBMatrix
+    translateHelper columns dim = case someNatVal (fromIntegral dim) of
+      SomeNat (_ :: Proxy dim) -> let result :: SBMatrix dim = SBMatrix (fromJust (SV.fromList (map toIndices columns))) in
+        SomeKnown result
+          where
+            toIndices :: KnownNat dim => SignedPrimeIntSet -> SBVector dim
+            toIndices x = SBVector $ U.fromList $ map fromIntegral $ if sign x then 0 : primeTranslation else primeTranslation
+                  where
+                    primeTranslation :: [Int]
+                    primeTranslation = binarySearch (PS.toAscList (primeSet x)) $ indexPrimes columns
+
+linearSolve' :: SomeKnown SBMatrix -> SomeKnown DBVector
+linearSolve' (SomeKnown m) = SomeKnown (linearSolve m)
+
+indexPrimes :: [SignedPrimeIntSet] -> U.Vector (Prime Int)
+indexPrimes list = U.fromList . PS.toAscList $ go list
+  where
+    go :: [SignedPrimeIntSet] -> PS.PrimeIntSet
+    go = foldr ((<>) . primeSet) mempty
+
+binarySearch :: (Eq a, Ord a, U.Unbox a) => [a] -> U.Vector a -> [Int]
+binarySearch list v = go 0 (len - 1) list v
+  where
+    len = U.length v
+    go :: (Eq a, Ord a, U.Unbox a) => Int -> Int -> [a] -> U.Vector a -> [Int]
+    go _ _ [] _ = []
+    go lowerIndex upperIndex allItems@(item : otherItems) vector = case item `compare` entry of
+      LT -> go lowerIndex (currentIndex - 1) allItems vector
+      -- @(currentIndex + 1)@ makes sure no prime number index is 0
+      -- 0 is reserved for negative numbers
```

Comment:
This is really unexpected semantics for a function, named `binarySearch`. If users of this function need indices starting from 1, they can apply `map (+1)` themselves.

Reply:
I added a comment to explain it better. Suppose I have Prime 2, Prime 3 and Prime 5. I want to send them respectively to 1,2,3. I use 0 to when the number is negative (its sign is true).

Reply:
To make it clearer `30` would be given by `[1,2,3]` whereas `-30` by `[0,1,2,3]`.